File: Unit4P.ts
/**
 * @author William Provost
 * @version 1.0.0
 * @date 2025-12-09
 * @fileoverview This program keeps track of car stats and simulates drive/fill/oil change
 */

// Initial data (per assignment)
const make: string = "UsedCarBrand";
const model: string = "ModelX";
let carColor: string = "Silver";
let odometer: number = 65000;       // starting mileage
let oilChangeKM: number = 65000;    // last oil change at 65000
const gasCost: number[] = new Array(10);
let fillIndex: number = 0;

// initial fill-up per assignment
gasCost[fillIndex++] = 74.0;

// Function: carStats() returns a string of car info
function carStats(): string {
  return (
    `Car Make: ${make}\n` +
    `Car Model: ${model}\n` +
    `Car Colour: ${carColor}\n` +
    `Odometer: ${odometer} km\n` +
    `Last oil change at: ${oilChangeKM} km\n` +
    `Number of recorded fill-ups: ${fillIndex}`
  );
}

// Function: wrapCar() prompts for a new colour and returns it
function wrapCar(): string {
  const newColour: string = prompt("Enter a new colour to wrap the car with: ") || carColor;
  return newColour.trim();
}

// Function: drive() returns random km driven and updates odometer
function drive(): number {
  const driven: number = Math.floor(Math.random() * 901) + 100; // 100..1000
  odometer += driven; // update odometer internally
  return driven;
}

// Function: fillUp() prompts user for the price paid and stores it in gasCost
function fillUp(): void {
  if (fillIndex >= gasCost.length) {
    console.log("Gas cost array is full; cannot record more fill-ups.");
    return;
  }
  const input = prompt("Enter how much you paid for the gas to fill up: ") || "0";
  const cost = Number(input);
  gasCost[fillIndex++] = cost;
  console.log(`Recorded fill-up of $${cost.toFixed(2)} at slot ${fillIndex - 1}.`);
}

// Function: displayCostToFillUp() shows all recorded costs and returns average
function displayCostToFillUp(): number {
  if (fillIndex === 0) {
    console.log("No fill-up records.");
    return 0;
  }
  console.log("\nRecorded fill-up costs:");
  let sum = 0;
  for (let i = 0; i < fillIndex; i++) {
    console.log(`Fill-up ${i + 1}: $${gasCost[i].toFixed(2)}`);
    sum += gasCost[i];
  }
  const avg = sum / fillIndex;
  console.log(`Average cost per fill-up: $${avg.toFixed(2)}`);
  return avg;
}

// Function: oilChange() returns boolean, prints message, and updates oilChangeKM internally
function oilChange(): boolean {
  if (odometer - oilChangeKM >= 5000) {
    oilChangeKM = odometer; // update last oil change
    console.log("\nAn oil change was done.");
    return true;
  } else {
    console.log("\nYour car does not need an oil change.");
    return false;
  }
}

/* ----------------- Main program ----------------- */

console.log("Initial car stats:");
console.log(carStats());

// wrap the car (change colour)
const colourInput = wrapCar();
carColor = colourInput;
console.log(`\nCar colour updated to: ${carColor}`);

// drive
const drivenKm = drive();
console.log(`\nYou drove ${drivenKm} km. New odometer: ${odometer} km.`);

// ask user to fill up now
fillUp();

// display fill-up costs and average
const avgCost = displayCostToFillUp();

// check oil change
oilChange();

console.log("\nFinal car stats:");
console.log(carStats());

console.log("\nDone.");

File: Unit4P.go
// Author: William Provost
// Version: 1.0.0
// Date: 2025-12-10
// Fileoverview: This program keeps track of car stats and simulates drive/fill/oil change

package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Initial data
var makeCar string = "UsedCarBrand"
var model string = "ModelX"
var carColor string = "Silver"
var odometer int = 65000    // starting mileage
var oilChangeKM int = 65000 // last oil change
var gasCost [10]float64
var fillIndex int = 0

// carStats returns a string of car info
func carStats() string {
	return fmt.Sprintf(
		"Car Make: %s\nCar Model: %s\nCar Colour: %s\nOdometer: %d km\nLast oil change at: %d km\nNumber of recorded fill-ups: %d",
		makeCar, model, carColor, odometer, oilChangeKM, fillIndex)
}

// wrapCar prompts user for new colour
func wrapCar() string {
	var newColour string
	fmt.Print("Enter a new colour to wrap the car with: ")
	fmt.Scanln(&newColour)
	if newColour == "" {
		return carColor
	}
	return newColour
}

// drive generates random km (100..1000) and updates odometer
func drive() int {
	rand.Seed(time.Now().UnixNano())
	driven := rand.Intn(901) + 100 // 100..1000
	odometer += driven
	return driven
}

// fillUp prompts for cost and stores in gasCost array
func fillUp() {
	if fillIndex >= len(gasCost) {
		fmt.Println("Gas cost array is full; cannot record more fill-ups.")
		return
	}
	var cost float64
	fmt.Print("Enter how much you paid for the gas to fill up: ")
	fmt.Scanln(&cost)
	gasCost[fillIndex] = cost
	fmt.Printf("Recorded fill-up of $%.2f at slot %d.\n", cost, fillIndex)
	fillIndex++
}

// displayCostToFillUp shows all recorded costs and returns average
func displayCostToFillUp() float64 {
	if fillIndex == 0 {
		fmt.Println("No fill-up records.")
		return 0
	}
	fmt.Println("\nRecorded fill-up costs:")
	sum := 0.0
	for i := 0; i < fillIndex; i++ {
		fmt.Printf("Fill-up %d: $%.2f\n", i+1, gasCost[i])
		sum += gasCost[i]
	}
	avg := sum / float64(fillIndex)
	fmt.Printf("Average cost per fill-up: $%.2f\n", avg)
	fmt.Printf("\n(Info) Returned average cost: $%.2f\n", avg)
	return avg
}

// oilChange checks if vehicle needs oil change and updates oilChangeKM
func oilChange() bool {
	if odometer-oilChangeKM >= 5000 {
		oilChangeKM = odometer
		fmt.Println("\nAn oil change was done.")
		return true
	}
	fmt.Println("\nYour car does not need an oil change.")
	return false
}

func main() {
	// initial fill-up 
	gasCost[fillIndex] = 74.0
	fillIndex++

	fmt.Println("Initial car stats:")
	fmt.Println(carStats())

	// wrap the car
	carColor = wrapCar()
	fmt.Printf("\nCar colour updated to: %s\n", carColor)

	// drive
	drivenKm := drive()
	fmt.Printf("\nYou drove %d km. New odometer: %d km.\n", drivenKm, odometer)

	// fill up
	fillUp()

	// display fill-up costs and average
	_ = displayCostToFillUp()

	// check oil change
	oilChange()

	fmt.Println("\nFinal car stats:")
	fmt.Println(carStats()) // <-- this shows odometer, oilChangeKM, and fillIndex

	fmt.Println("\nDone.")
}
